lib/Conversion/RTLToLLHD/RTLToLLHD.cpp

```diff
@@ -0,0 +1,178 @@
+//===- RTLToLLHD.cpp - RTL to LLHD Conversion Pass ------------------------===//
+//
+// This is the main RTL to LLHD Conversion Pass Implementation.
+//
+//===----------------------------------------------------------------------===//
+
+#include "circt/Conversion/RTLToLLHD/RTLToLLHD.h"
+#include "circt/Dialect/LLHD/IR/LLHDDialect.h"
+#include "circt/Dialect/LLHD/IR/LLHDOps.h"
+#include "circt/Dialect/RTL/Dialect.h"
+#include "circt/Dialect/RTL/Ops.h"
+#include "mlir/IR/StandardTypes.h"
+#include "mlir/Pass/Pass.h"
+#include "mlir/Transforms/DialectConversion.h"
+
+namespace circt {
+namespace llhd {
+#define GEN_PASS_CLASSES
+#include "circt/Conversion/RTLToLLHD/Passes.h.inc"
+} // namespace llhd
+} // namespace circt
+
+using namespace circt;
+using namespace llhd;
+using namespace rtl;
+
+//===----------------------------------------------------------------------===//
+// RTL to LLHD Conversion Pass
+//===----------------------------------------------------------------------===//
+
+namespace {
+struct RTLToLLHDPass : public ConvertRTLToLLHDBase<RTLToLLHDPass> {
+  void runOnOperation() override;
+};
+
+/// A helper type converter class that automatically populates the relevant
+/// materializations and type conversions for converting RTL to LLHD.
+struct RTLToLLHDTypeConverter : public TypeConverter {
+  RTLToLLHDTypeConverter();
+};
+} // namespace
+
+/// Create a RTL to LLHD conversion pass.
+std::unique_ptr<OperationPass<mlir::ModuleOp>>
+llhd::createConvertRTLToLLHDPass() {
+  return std::make_unique<RTLToLLHDPass>();
+}
+
+/// Register the RTL to LLHD conversion pass.
+namespace {
+#define GEN_PASS_REGISTRATION
+#include "circt/Conversion/RTLToLLHD/Passes.h.inc"
+} // namespace
+
+void circt::llhd::registerRTLToLLHDPasses() { registerPasses(); }
+
+/// Forward declare conversion patterns.
+struct ConvertRTLModule;
+struct ConvertOutput;
+
+/// This is the main entrypoint for the RTL to LLHD conversion pass.
+void RTLToLLHDPass::runOnOperation() {
+  MLIRContext &context = getContext();
+  ModuleOp module = getOperation();
+
+  ConversionTarget target(context);
+  target.addLegalDialect<LLHDDialect>();
+  target.addIllegalOp<RTLModuleOp>();
+
+  RTLToLLHDTypeConverter typeConverter;
+  OwningRewritePatternList patterns;
+  patterns.insert<ConvertRTLModule>(typeConverter, &context);
+  patterns.insert<ConvertOutput>(typeConverter, &context);
+
+  if (failed(applyPartialConversion(module, target, std::move(patterns))))
+    signalPassFailure();
+}
+
+//===----------------------------------------------------------------------===//
+// TypeConverter conversions and materializations
+//===----------------------------------------------------------------------===//
+RTLToLLHDTypeConverter::RTLToLLHDTypeConverter() {
+  // Convert IntegerType by just wrapping it in SigType.
+  addConversion([](IntegerType type) { return SigType::get(type); });
```

Comment:
This is good start in my opinion. It's not quite clear to me yet how we'll decide to convert something in the RTL dialect to a signal with associated `prb` and `drv`, and when to chain multiple RTL ops together without writing back to a signal. Maybe there is a sort of "signal-to-reg" pass hidden in there somewhere, alongside LLVM's mem-to-reg 🤔.

Reply:
If we can frame the problem as type conversion, the dialect conversion infrastructure provides some nice helpers to materialize casts. In this case, we could materialize a signal with associated `prb` and `drv` ops as the casts.

I'm not sure how far we can take that with respect to chaining together multiple ops without writing back to a signal, but the cast materialization approach works nicely if it is a good fit.